lib/erblint-github/linters/tag_tree_helpers.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module ERBLint
+  module Linters
+    # Helpers used by linters to organize HTML tags into abstract syntax trees.
+    module TagTreeHelpers
+      # from https://github.com/Shopify/erb-lint/blob/6179ee2d9d681a6ec4dd02351a1e30eefa748d3d/lib/erb_lint/linters/self_closing_tag.rb
+      SELF_CLOSING_TAGS = %w[
+        area base br col command embed hr input keygen
+        link menuitem meta param source track wbr img
+      ].freeze
+
+      # This assumes that the AST provided represents valid HTML, where each tag has a corresponding closing tag.
+      # From the tags, we build a structured tree which represents the tag hierarchy.
+      # With this, we are able to know where the tags start and end.
+      def build_tag_tree(processed_source)
```

Comment:
This is awesome!!  

> This assumes that the AST provided represents valid HTML

What happens when the AST doesn't represent valid HTML? I'm guessing ERB lint rules like `SelfClosingTag` and `ParserErrors` ensures we don't end up in that state? 

Should we add tests for this method, either in this PR or as a follow up?